qiskit/quantum_info/synthesis/one_qubit_decompose.py

```diff
@@ -416,32 +422,46 @@ def _circuit_psx_gen(theta, phi, lam, phase, atol, pfun, xfun, xpifun=None):
         """Generic X90, phase decomposition"""
         qr = QuantumRegister(1, "qr")
         circuit = QuantumCircuit(qr, global_phase=phase)
-        # Check for decomposition into minimimal number required SX pulses
+        # Early return for zero SX decomposition
         if np.abs(theta) < atol:
-            # Zero SX gate decomposition
-            pfun(circuit, qr, lam + phi)
+            if abs(_mod_2pi(lam + phi)) > atol:
+                pfun(circuit, qr, lam + phi)
             return circuit
+        # Early return for single SX decomposition
         if abs(theta - np.pi / 2) < atol:
-            # Single SX gate decomposition
-            pfun(circuit, qr, lam - np.pi / 2)
+            if abs(_mod_2pi(lam - np.pi / 2)) > atol:
+                pfun(circuit, qr, lam - np.pi / 2)
             xfun(circuit, qr)
-            pfun(circuit, qr, phi + np.pi / 2)
+            if abs(_mod_2pi(phi + np.pi / 2)) > atol:
+                pfun(circuit, qr, phi + np.pi / 2)
             return circuit
-        # General two-SX gate decomposition
-        # Shift theta and phi so decomposition is
-        # P(phi).SX.P(theta).SX.P(lam)
+        # General double SX decomposition
         if abs(theta - np.pi) < atol:
             circuit.global_phase += lam
             phi, lam = phi - lam, 0
+        if abs(_mod_2pi(lam + np.pi)) < atol:
+            lam, theta, phi = 0, -theta, phi + lam
+            circuit.global_phase -= theta
+        if abs(_mod_2pi(phi)) < atol:
+            lam, theta, phi = lam - np.pi, -theta, phi + np.pi
+            circuit.global_phase -= theta
```

Comment:
maybe clearer to move this after the shift? Probably also better to have a single if:
```python
if abs(_mod_2pi(phi-pi))<atol or abs(_mod_2pi(lam-pi))<atol:
    theta, phi, lam = -theta, phi+pi, lam+pi
```

Reply:
Agreed about the `if`. I found this clearer before the shift: I spent a while confused about why these extra RZ(pi)s weren't flipping the SXes to SX^*es, but when that chunk is written as an RY I see it straight away.